ledger/voters.go

```diff
@@ -0,0 +1,337 @@
+
+	if proto.CompactCertRounds == 0 {
+		// Disabled, nothing to load.
+		return nil
+	}
+
+	// Walk backwards to find blocks R such that R+CompactCertVotersLookback
+	// is a multiple of CompactCertRounds, and where we do not have a complete
+	// compact cert yet (hdr.CompactCertLastRound).
+	rPlusLookback := basics.Round(((uint64(latest) + proto.CompactCertVotersLookback) / proto.CompactCertRounds) * proto.CompactCertRounds)
+	minRound := basics.Round(proto.CompactCertVotersLookback)
+	if hdr.CompactCertLastRound >= minRound {
+		minRound = hdr.CompactCertLastRound
+	}
+	for rPlusLookback >= minRound {
+		r := rPlusLookback.SubSaturate(basics.Round(proto.CompactCertVotersLookback))
+		hdr, err = l.BlockHdr(r)
+		if err != nil {
+			switch err.(type) {
+			case ErrNoEntry:
+				// If we cannot retrieve a block to construct the tree
+				// then we must have already evicted that block, which
+				// must have been because compact certs weren't enabled
+				// when that block was being considered for eviction.
+				// OK to stop.
+				return nil
+			default:
+				return err
+			}
+		}
+
+		if config.Consensus[hdr.CurrentProtocol].CompactCertRounds == 0 {
+			// Walked backwards past a protocol upgrade, no more compact certs.
+			return nil
+		}
+
+		err = vt.loadTree(hdr)
+		if err != nil {
+			return err
+		}
+
+		rPlusLookback = rPlusLookback.SubSaturate(basics.Round(proto.CompactCertRounds))
+	}
+
+	return nil
+}
+
+func (vt *votersTracker) loadTree(hdr bookkeeping.BlockHeader) error {
+	r := hdr.Round
+
+	_, ok := vt.round[r]
+	if ok {
+		// Already loaded.
+		return nil
+	}
+
+	proto := config.Consensus[hdr.CurrentProtocol]
+	if proto.CompactCertRounds == 0 {
+		// No compact certs.
+		return nil
+	}
+
+	tr := &VotersForRound{
+		Proto: proto,
+	}
+	tr.cond = sync.NewCond(&tr.mu)
+	vt.round[r] = tr
+
+	go func() {
+		err := tr.loadTree(vt.l, vt.au, hdr)
+		if err != nil {
+			vt.au.log.Warnf("votersTracker.loadTree(%d): %v", hdr.Round, err)
+		}
+	}()
+	return nil
+}
+
+func (tr *VotersForRound) loadTree(l ledgerForTracker, au *accountUpdates, hdr bookkeeping.BlockHeader) error {
+	r := hdr.Round
+
+	// certRound is the block that we expect to form a compact certificate for,
+	// using the balances from round r.
+	certRound := r + basics.Round(tr.Proto.CompactCertVotersLookback+tr.Proto.CompactCertRounds)
+
+	// sigKeyRound is the ephemeral key ID that we expect to be used for signing
+	// the block from certRound.  It is one higher because the keys for certRound
+	// might be deleted by the time consensus is reached on the block and we try
+	// to sign the compact cert for block certRound.
+	sigKeyRound := certRound + 1
+
+	top, err := au.onlineTop(r, sigKeyRound, tr.Proto.CompactCertTopVoters)
+	if err != nil {
+		return err
+	}
+
+	participants := make(participantsArray, len(top))
+	addrToPos := make(map[basics.Address]uint64)
+	var totalWeight basics.MicroAlgos
+
+	for i, acct := range top {
+		var ot basics.OverflowTracker
```

Comment:
nit: move the `basics.OverflowTracker` outside the for loop.

Reply:
You're right that we could move it out (since, on each iteration of the loop, either `ot` remains non-overflowed, or the loop returns an error).  But I think that would make the code harder to read, because `ot` is used independently for each iteration of the loop, and it's not used after the loop.